File: symfony/src/Service/Players/CraftingRecipeService.php
<?php

namespace App\Service\Players;

use App\Entity\Players\CraftingRecipe;
use App\Repository\Players\CraftingRecipeRepository;

class CraftingRecipeService
{
    public function __construct(
        private CraftingRecipeRepository $repository,
    ) {}

    public function create(array $data): CraftingRecipe
    {
        throw new \LogicException('Not implemented');
    }

    public function update(CraftingRecipe $entity, array $data): CraftingRecipe
    {
        throw new \LogicException('Not implemented');
    }

    public function canCraft(int $id, $playerId): mixed
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('CraftingRecipe not found: ' . $id);
        $result = $entity->canCraft($playerId);
        $this->repository->save($entity, flush: true);
        return $result;
    }

    public function executeCraft(int $id, $playerId): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('CraftingRecipe not found: ' . $id);
        $entity->executeCraft($playerId);
        $this->repository->save($entity, flush: true);
    }

    public function disable(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('CraftingRecipe not found: ' . $id);
        $entity->disable();
        $this->repository->save($entity, flush: true);
    }

    public function enable(int $id): void
    {
        $entity = $this->repository->find($id);
        if (!$entity) throw new \RuntimeException('CraftingRecipe not found: ' . $id);
        $entity->enable();
        $this->repository->save($entity, flush: true);
    }
}

File: rails/app/services/players/crafting_recipe_service.rb
module Players
  class CraftingRecipeService
    def initialize(repository = CraftingRecipe)
      @repository = repository
    end

    def create(attributes)
      raise NotImplementedError, "#{self.class}#create not implemented"
    end

    def update(entity, attributes)
      raise NotImplementedError, "#{self.class}#update not implemented"
    end

    def can_craft(id)
      instance = CraftingRecipe.find(id)
      result = instance.can_craft(player_id)
      instance.save!
      result
    end

    def execute_craft(id, player_id)
      instance = CraftingRecipe.find(id)
      instance.execute_craft(player_id)
      instance.save!
    end

    def disable(id)
      instance = CraftingRecipe.find(id)
      instance.disable()
      instance.save!
    end

    def enable(id)
      instance = CraftingRecipe.find(id)
      instance.enable()
      instance.save!
    end
  end
end
